lib/task.ml

```diff
@@ -0,0 +1,83 @@
+type 'a task = unit -> 'a
+
+type 'a promise = ('a, exn) result option Atomic.t
+
+exception TasksActive of Domain.id
+
+type task_msg =
+  Task : 'a task * 'a promise -> task_msg
+| Quit : task_msg
+
+type pool =
+  {domains : unit Domain.t array;
+   task_chan : task_msg Chan.t}
+
+let do_task f p =
+  try
+    let res = f () in
+    Atomic.set p (Some (Ok res))
+  with e ->
+    Atomic.set p (Some (Error e));
+    match e with
+    | TasksActive _ -> raise e
+    | _ -> ()
+
+let setup_pool ~num_domains =
+  let task_chan = Chan.make_unbounded () in
+  let rec worker () =
+    match Chan.recv task_chan with
+    | Quit -> ()
+    | Task (t, p) ->
+        do_task t p;
+        worker ()
+  in
+  let domains = Array.init num_domains (fun _ -> Domain.spawn worker) in
+  {domains; task_chan}
+
+let async pool task =
+  let p = Atomic.make None in
+  Chan.send pool.task_chan (Task(task,p));
+  p
+
+let rec await pool promise =
+  match Atomic.get promise with
+  | None ->
+      begin match Chan.recv_poll pool.task_chan with
+      | None -> Domain.Sync.cpu_relax ()
+      | Some (Task (t, p)) -> do_task t p
+      | Some Quit -> raise (TasksActive (Domain.self ()))
```

Comment:
I removed the `Domain.id` argument from `TasksActive` since that information is not useful (and incorrect). 

The correct use of `Tasks` will need to ensure that all the promises created are complete before the pool is torn down. If not, `teardown_pool` will raise `TasksActive` exception. I've also added an example in tests (`task_exn.ml`) to show exceptional control flow when using `Tasks`.